### ISSUE TITLE
Cannot create multiple path-pattern conditions for ALB Rules

### ISSUE THREAD

**Issue by emmm-dee** [terraform-providers/terraform-provider-aws#8540]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version
```
Terraform v0.11.13
+ provider.aws v2.8.0
+ provider.template v2.1.1
```

### Affected Resource(s)

* aws_lb_listener_rule

### Terraform Configuration Files

```hcl
resource "aws_lb_listener_rule" "test" {
  listener_arn = "<<arn_scrubbed>>"
  priority     = 25
  action {
    type             = "forward"
    target_group_arn = "${aws_lb_target_group.mytarget.arn}"
  }

  condition {
    field  = "path-pattern"
    values = ["/account.php", "/client.php*"]
  }

  condition {
    field  = "host-header"
    values = ["dev01site.example.com"]
  }
}
```

### Expected Behavior

Created an ALB rule with a multi-condition path-pattern - this is possible to do in the console. 

See screenshot:
![Screen Shot 2019-05-06 at 10 42 30 PM](https://user-images.githubusercontent.com/45244869/57275878-a2a81900-7054-11e9-8f43-fc5c0f56e7e0.png)

### Actual Behavior

`Error: aws_lb_listener_rule.test: condition.0.values: attribute supports 1 item maximum, config has 2 declared`

This is the first time I'm seeing an option in the Amazon Console GUI that I cannot perform with Terraform. Have triple checked the TF docs for aws_lb_listener_rule and it even states "A maximum of 1 can be defined." for the Values field.

**Comment by sethbacon** [comment]
Also host-header is not supporting multiple values.

Error: aws_alb_listener_rule.alb_web_rule1: condition.0.values: attribute supports 1 item maximum, config has 2 declared.

resource "aws_alb_listener_rule" "alb_web_rule1" {
  listener_arn = "${aws_alb_listener.alb_web_https.arn}"
  priority     = 1
  action {
    type             = "forward"
    target_group_arn = "${aws_alb_target_group.tg_web1_uat_https.arn}"
  }
  condition {
    field  = "host-header"
    values = ["*uat.example.com","*uat-analytics.example.com"]
  }
}

**Comment by agrzegorczyk-leonsoftware** [comment]
#8126

**Comment by thoo5ieb** [comment]
When I use resource `aws_lb_listener_rule` with multiple path patterns, I follow code practice below:

```
variable "path_patterns" {
  type = "list"
}

resource "aws_lb_listener_rule" "default" {
  count = "${length(var.path_patterns)}"
  listener_arn = "${aws_lb_listener.default.arn}"
  
  /* .. */
  
  condition {
    field = "path-pattern"
    
    values = [
      "${element(var.path_patterns, count.index)}",
    ]
  }
}
```

Generally speaking, I attach multiple `aws_lb_listener_rule` to `aws_lb_listener`.

**Comment by sandangel** [comment]
I hit the ALB limit number of rules with the solution that @thoo5ieb provided. Is this feature available in provider version 2.23.0?

**Comment by thoo5ieb** [comment]
@sandangel If I understand correctly, the limit reaching issue you have, will be pretty much the same with different Terraform resource design constraints. Under the hood, it still does N rules for a specific AWS ALB listener. One way or other, you can reach out AWS support OR re-design (if possible) your solution to a given problem.

**Comment by sandangel** [comment]
@thoo5ieb Thanks for your response.
I have around 107 routes and With your solution I need to use 107 rules for https listener, which is not possible with the current ALB spec (100 rules max). 
With multiple values per rule config I just need to use 22 rules, which is still far from the limit. I managed to reduced the number of routes to  < 100 by using wildcard prefix for some routes so your solution work for me right now. but hopefully multiple values per rule will available soon.

**Comment by emmm-dee** [comment]
@thoo5ieb,  @sandangel hit the nail on the head and explained the exact use case on why I created this issue in the first place. 

The functionality (multiple pathing rules) can be done with the AWS console, but cannot be done with Terraform. Therefore if we want to not hit the limit, we cannot use Terraform and need to use the AWS console or some janky wildcard workaround which doesn't fit all use-cases.

**Comment by sandangel** [comment]
@emmm-dee you can still use local_exec or null_resource to call aws-cli

**Comment by emmm-dee** [comment]
> @emmm-dee you can still use local_exec , null_resource to call aws-cli

For sure, but quite sad it's the only way given that all my other infra is properly managed with Terraform. 

Right now I'm just using a bash script that calls aws-cli to get this job done. Could probably have local_exec call it, but it's just as easy triggering the script as needed manually.

**Comment by scalp42** [comment]
@sandangel any chance you could share an example?

**Comment by mcastellini** [comment]
This issue is part of #8126

**Comment by thabiger** [comment]
Here is a working workaround example from my code with few explanations:

- listener rules made by Terraform contain single pattern value(as condition block is required and rule resource won't create without it) and is then overwritten by the null_resource provider later in plan execution
- on each subsequent run, Terraform will find that the actual rule state doesn't match the desired one and restore configuration managed by the aws_alb_listener_rule resource
- null_resource is not aware of actual state as it only executes provisioners; in particular, it won't know terraform restored it's desired config, so it must be triggered on each Terraform's run with timestamp() value
- the overall result is rules are recreated on each Terraform's run (which is not great but works, allows keeping everything in terraform code and prevents from hitting ALB limits due to excessive rules separation)

```
resource "aws_alb_listener_rule" " example_https" {
  listener_arn = "${data.aws_lb_listener.elb_service_listener_443.arn}"
  action {
    type             = "forward"
    target_group_arn = "${module.elb_example_targetgroup.arn}"
  }
  condition {
    field  = "host-header"
    values = ["example.com"]
  }
}
  resource "null_resource" "update_rule_example_https" {
    triggers = {
      always = "${timestamp()}"
    }
    provisioner "local-exec" {
      command = <<EOF
        aws elbv2 modify-rule \
          --rule-arn=${aws_alb_listener_rule.example_https.arn} \
          --conditions='[
    {
        "Field": "host-header",
        "Values": [
            "example.com",
            "www.example.com"
        ]
    }
]'
EOF
    }
  }
```

**Comment by danieladams456** [comment]
It looks like this is possible with the new syntax:
```hcl
  condition {
    host_header {
      values = var.alb_listener_rule_host_header_values
    }
  }
  condition {
    path_pattern {
      values = var.alb_listener_rule_path_pattern_values
    }
  }
```

It also works with terraform 0.12 dynamic blocks:
```hcl
resource "aws_alb_listener_rule" "service" {
  listener_arn = data.aws_alb_listener.service.arn

  action {
    type             = "forward"
    target_group_arn = aws_alb_target_group.service.arn
  }

  condition {
    host_header {
      values = var.alb_listener_rule_host_header_values
    }
  }
  
  dynamic "condition" {
    for_each = length(var.alb_listener_rule_path_pattern_values) > 0 ? [true] : []
    content {
      path_pattern {
        values = var.alb_listener_rule_path_pattern_values
      }
    }
  }
}
```

**Comment by bflad** [comment]
Hi folks 👋 Sorry this issue was not closed out with the release of version 2.42.0 of the Terraform AWS Provider a few weeks ago as part of https://github.com/terraform-providers/terraform-provider-aws/issues/8126. Closing as the feature request was resolved.

The examples provided by @danieladams456 above should be a good starting point. If you're looking for general assistance still, please note that we use GitHub issues in this repository for tracking bugs and enhancements with the Terraform AWS Provider codebase rather than for questions. While we may be able to help with certain simple problems here it's generally better to use the [community forums](https://discuss.hashicorp.com/c/terraform-providers) where there are far more people ready to help, whereas the GitHub issues here are generally monitored only by a few maintainers and dedicated community members interested in code development of the Terraform AWS Provider itself.
